Revamp how the cockroachdb/cockroach image is built.

petermattis opened the pull request:
We now use build/builder.sh for building the cockroach binary and tests
and run the new acceptance tests instead of run/local-cluster.sh. The
latter is now deleted.

Fixes #1754.

petermattis committed:
Revamp how the cockroachdb/cockroach image is built.

We now use build/builder.sh for building the cockroach binary and tests
and run the new acceptance tests instead of run/local-cluster.sh. The
latter is now deleted.

Fixes #1754.

petermattis commented on build/deploy/mkimage.sh:
```diff
@@ -12,26 +12,35 @@
+# commands in the if-branch to be executed within the docker
+# container.
+if [ "${1:-}" = "docker" ]; then
+    time make testbuild
```
Is building the tests statically like this worthwhile? It takes quite a while and seems duplicative of the tests that are run build `circle-test.sh`.

bdarnell replied:
I don't think so; even without the build time there's a non-trivial amount of complexity devoted to getting these tests built and running in a strange artificial environment.

bdarnell commented on build/deploy/cockroach.sh:
```diff
@@ -2,13 +2,13 @@
 if [ "${1:-}" = "shell" ]; then
   shift
-  /bin/bash "$@"
+  /bin/sh "$@"
```
Why change this to `sh`? We use `bash` in all our `#!` lines and use bashisms, well, unabashedly.

petermattis replied:
Oops, I was thinking of switching to use `busybox` as the base image, but reverted that. I'll revert this as well.

bdarnell commented on build/build-docker-deploy.sh:
```diff
@@ -2,7 +2,4 @@
-
-./build/build-docker-dev.sh
-./build/deploy/mkimage.sh
+$(dirname $0)/deploy/mkimage.sh
```
Why keep this file around if it's just a wrapper around mkimage.sh?

petermattis replied:
It's mentioned in the docs? Perhaps I should move the contents of `deploy/mkimage.sh` to this file.

petermattis replied:
This is done. I've moved the contents of `deploy/mkimage.sh` into `build/build-docker-deploy.sh`.

bdarnell commented on build/deploy/mkimage.sh:
```diff
@@ -12,26 +12,35 @@
 
-cd -P "$(dirname $0)"
-DIR=$(pwd -P)
+# This is mildly trick: This script runs itself recursively. The first
```
s/trick/tricky/

petermattis replied:
Fixed.

bdarnell commented:
LGTM

petermattis committed:
Further tweaks to cockroachdb/cockroach image building.

No longer build the tests and run them inside the deploy container. This
work was duplicative of the normal testing and served little
purpose. Removed the associated "testbuild" target from the Makefile.

Switched to using "busybox" as the base image. This reduces the
cockroachdb/cockroach image size from 210MB to 88MB.

petermattis commented:
PTAL at the new commit.

bdarnell commented:
LGTM

tschottdorf commented:
the reason we were running the tests again in the deploy image was that the static build can behave differently. For example, failure to correctly use `netgo` and putting the "static" binary into a `busybox` will cause all DNS lookups to fail. I don't think this warrants all the effort put into it previously, so as long as there's some acceptance test that makes sure the deploy image isn't completely toast, that'll do.

petermattis commented:
@tschottdorf Yep, I'm still running the acceptance tests. Curious about the '*.test' binaries, though. They weren't being built with `-tags netgo`, so the DNS behavior would be different from the cockroach binary.
